share/csm_share/shr/shr_func_mod.F90

```diff
@@ -0,0 +1,58 @@
+!===============================================================================
+! SVN $Id: shr_func_mod.F90 25434 2010-11-04 22:46:24Z tcraig $
+! SVN $URL: https://svn-ccsm-models.cgd.ucar.edu/csm_share/release_tags/cesm1_2_x_n02_share3_130715/shr/shr_func_mod.F90 $
+!===============================================================================
+
+MODULE shr_func_mod
+
+   use shr_kind_mod, only: SHR_KIND_R8, SHR_KIND_IN, SHR_KIND_CS
+   use shr_sys_mod
+   use shr_const_mod
+   use shr_log_mod, only: s_loglev  => shr_log_Level
+   use shr_log_mod, only: s_logunit => shr_log_Unit
+
+   IMPLICIT none
+
+   !----------------------------------------------------------------------------
+   ! PUBLIC: Interfaces and global data
+   !----------------------------------------------------------------------------
+   public :: shr_func_freezetemp
+
+   character(SHR_KIND_CS) :: tfreeze_option     ! option for computing freezing point
+                                    ! default is constant -1.8C
+                                    ! 'linear_salt' for linear equation
+                                    ! 'mushy' for CICE mushy-layer nonlinear equation
+
+!===============================================================================
+CONTAINS
+!===============================================================================
+
+real(SHR_KIND_R8) pure FUNCTION shr_func_freezetemp(s)
+
+   !----------------------------------------------------------------------------
+   !
+   ! FUNCTION to return the freezing point of salt water
+   !
+   !--------------- Code History -----------------------------------------------
+   !
+   ! Original Author: David Bailey
+   ! Date:            Feb, 2016
+   !----------------------------------------------------------------------------
+
+   real   (SHR_KIND_R8),intent(in) :: s ! Salinity in psu
+
+   !----------------------------------------------------------------------------
+
+   if (tfreeze_option == 'linear_salt') then
```

Comment:
It looks like this function will be called once for every point, each time through the time loop. If so, performance of this string comparison could matter on some platforms.

In general, I prefer that string options read from the namelist be converted to integers in initialization, like:

integer, parameter :: TFREEZE_OPTION_LINEAR_SALT = 1
integer, parameter :: TFREEZE_OPTION_MUSHY = 2
etc.

Then you can do integer compares rather than string compares.

Also, maybe more importantly: Doing comparisons against named constants would be safer here. For example, if someone changed the option names to be capitalized ('Linear_Salt', 'Mushy'), the current code would silently return -1 regardless of the option you chose. (Even if the code were changed as I note below, comparisons against named constants still helps in that you get a compile-time error rather than a run-time error.)